### ISSUE TITLE
New feature: Tag classes should be accounted when writing to DER

### ISSUE THREAD

**Issue by ghost** [alex/rust-asn1#365]
Hello again,

thanks for your help so far! :)

I wanted to encode an asn1 specifiction, that uses `application` tags (also given implicit), i.e.:

```
InitiateSession ::= [APPLICATION 0] IMPLICIT SEQUENCE {
    a  [APPLICATION 10] IMPLICIT OCTET STRING (SIZE(8..32)),
    b  [APPLICATION 11] IMPLICIT OCTET STRING,
    c  [APPLICATION 31] IMPLICIT UTF8String (SIZE(1..512)) OPTIONAL,
    d  [APPLICATION 32] IMPLICIT OCTET STRING (SIZE(2..2)) OPTIONAL
}
```

and my first try was to use this struct with given derive macros:
```Rust

#[derive(asn1::Asn1Read, asn1::Asn1Write, PartialEq, Debug)]
struct InitiateSession<'a> {
    #[implicit(10, required)]
    a: &'a [u8],
    #[implicit(11, required)]
    b: &'a [u8],
    #[implicit(31)]
    c: Option<asn1::Utf8String<'a>>,
    #[implicit(32)]
    d: Option<&'a [u8]>,
};

let sid_a: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8];
let login_protocol_list: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8];
let system_use_text = "Hallo Yasin";
let server_msg_buf_size: Vec<u8> = vec![1, 2];

let session = InitiateSession {
    a: &sid_a,
    b: &login_protocol_list,
    c: Some(Utf8String::new(system_use_text)),
    d: Some(&server_msg_buf_size),
};

let bytes = asn1::write_single(&session).unwrap();
```

What I get are these bytes:
`30 27 8a 08 01 02 03 04 05 06 07 08 8b 08 01 02 03 04 05 06 07 08 9f 1f 0b 48 61 6c 6c 6f 20 59 61 73 69 6e 9f 20 02 01 02`

What I expect them to be (given with [asn1tools](https://pypi.org/project/asn1tools/))
`60 27 4a 08 01 02 03 04 05 06 07 08 4b 08 01 02 03 04 05 06 07 08 5f 1f 0b 48 61 6c 6c 6f 20 59 61 73 69 6e 5f 20 02 01 02`

What I am missing is the definition of [tag classes](https://letsencrypt.org/docs/a-warm-welcome-to-asn1-and-der/#tag-classes) when encoding. The above structure seems to use `universal` tags, as `0x30` is the encoding for the sequence (I think) in universal tags.

TLDR: Is there already a way to define the tag classes, when you want to encode?

**Comment by alex** [comment]
I'm not sure I understand the question, can you please try rephrasing it?

On Tue, Feb 21, 2023 at 11:31 AM YasinGuenduez-EH
***@***.***> wrote:
>
> Hello again,
>
> thanks for your help so far.
> I found tag classes that get constructed directly from the tag byte when parsing.
> I don't know if this feature already exists - or I just did not find it. Is there already a way to consider the tag class when writing to DER?
>
> If this feature already exists - then I would make a PR with a little example on how to do that - because I did not find it :).
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>



-- 
All that is necessary for evil to succeed is for good people to do nothing.

**Comment by YasinGuenduez-EH** [comment]
Hey Alex! Thanks for your immediate response! ?? I updated my issue on github.


Best regards 

Yasin Gündüz 


Software Engineer | TDG Technology Gateways

Endress+Hauser Digital Solutions

Endress+Hauser Process Solutions AG | Christoph Merian-Ring 4 | 4153 Reinach | Switzerland
Phone: +41 61 715 7319 | Fax: +41 61 715 2800
***@***.*** |  www.ehds.endress.com 




Disclaimer: 

The information transmitted is intended only for the person or entity to which it is addressed and may contain confidential, proprietary, and/or privileged material. Any review, retransmission, dissemination or other use of, or taking of any action in reliance upon, this information by persons or entities other than the intended recipient is prohibited. If you receive this in error, please contact the sender and delete the material from any computer. This e-mail does not constitute a contract offer, a contract amendment, or an acceptance of a contract offer unless explicitly and conspicuously designated or stated as such.

**Comment by alex** [comment]
There's not currently an API for tag classes.

**Comment by YasinGuenduez-EH** [comment]
Any plans to implement that in the future? I could have a look at it.

**Comment by alex** [comment]
To be more precise: the `Tag` class currently supports Tag Classes. But there's no way to marka field with a different tag class, besides by creating your own wrapper time.

I don't have any plans around this, but I'd be happy to discuss a proposed API for this.

**Comment by YasinGuenduez-EH** [comment]
In the current API implicit or explicit Tags are set. And within those, the tags are e.g. set to Context-Specific. You also define tags within the implementations of your `SimpleAsn1Writeable` traits. This is what I understand. 
The tags are already correctly encoded, when der Tag is of class e.g. `TagClass::Application` - that is what I've seen in the unit tests. 

So one has to define the tag class from outside. I don't know if this is the right place. But one could use e.g.:

```Rust
pub fn write_implicit_element_with_tagclass<T: Asn1Writable>(&mut self, val: &T, tag: u32, tag_class: TagClass) -> WriteResult {
    // Create an implicit tag with application tag class here...
    let tag = ...
   
    self.write_tlv(tag, |dest| Writer::new(dest).write_element(val))
}
```
and with an alias of `write_element::<Implicit<T, tag, tagclass>>`.
And if you do not supply the `tagclass`, it will default to the old behaviour. Just my thoughts.

**Comment by alex** [comment]
When tag class is used -- e.g. `[APPLICATION 10] OCTET STRING`, does the `[APPLICATION 10]` replace the `OCTET STRING` tag, or does it wrap it (the way `EXPLICIT` does), or do you always have to use `[APPLICATION]` WITH `EXPLICIT`/`IMPLICIT`?

**Comment by YasinGuenduez-EH** [comment]
[Letsencrypt](https://letsencrypt.org/docs/a-warm-welcome-to-asn1-and-der/#explicit-vs-implicit) states, that if you leave out the `IMPLICIT` or `EXPLICIT` tag, it defaults to the `EXPLICIT` one.

_"When the IMPLICIT or EXPLICIT keyword is not present, the default is EXPLICIT, unless the module sets a different default at the top with “EXPLICIT TAGS,” “IMPLICIT TAGS,” or “AUTOMATIC TAGS._”

So my assumption is, regarding your example, `[APPLICATION 10] OCTET STRING` should behave like `[APPLICATION 10] EXPLICIT OCTET STRING`.

**Comment by alex** [comment]
Got it, thanks for doing the research.

In this case, IMO, the _ideal_ API would be for `Explicit`/`Implicit` to be generic over `const TagClass`. However, const-generics don't yet support user defined types, so this doesn't work.

https://github.com/rust-lang/rust/issues/95174 is the tracking issue for that feature.

**Comment by YasinGuenduez-EH** [comment]
As I had kind of the same idea with the alias, I would use a primary type, i.e. `const u8` for the `const TagClass` as a placeholder and try it out that way. This would be sufficient for my use case for now with the fork. 
When const-generics with custom types are supported, I would reimplement that correctly with the enum and reopen this issue.

For now I'd close it. Thanks for helping out!

**Comment by YasinGuenduez-EH** [comment]
Will be reopened as soon as [user defined const parameters](https://github.com/rust-lang/rust/issues/95174) are supported.

**Comment by alex** [comment]
Let's keep this open, it's still a rela issue.

**Comment by yguenduez** [comment]
Ok, let me know if you have other ideas for the API. I will try the placeholder `const u8` for now.
